### PR TITLE
Build docker images when building packages

### PR DESCRIPTION
**Pull request by jsoriano**
Include generation of docker images in the process of building the rest of released packages by extending `mage package`. Ideally it could be extended to build images for windows too.

Dockerfile, entry point templates and default beats configurations based on [current ones](https://github.com/elastic/beats-docker/tree/master/templates).

Pending:
- [x] Fix permission issues
- [x] Export built image as artifact with `docker save` or similar
- [x] Complete parametrization of docker options
- [x] Make it work for all beats
  - [x] Support capabilities for packetbeat and heartbeat
  - [x] Run as root for auditbeat
  - [x] Heartbeat user is `heartbeat-elastic` :thinking: it should be `heartbeat` as other beats
  - [x] ~~Install libsystemd in journalbeat~~ (systemd-libs already installed in centos:7 base image)
  - [x] Support community beats
- [x] Check that beats can be configured using environment variables
- [x] Fix license label for licensed builds
- [x] Add packaging tests for docker

### CODE REVIEW

File: dev-tools/packaging/templates/docker/Dockerfile.tmpl [hunk 1]
```diff
@@ -0,0 +1,40 @@
+{{- $beatHome := printf "%s/%s" "/usr/share" .BeatName }}
+{{- $beatBinary := printf "%s/%s" $beatHome .BeatName }}
+
+FROM {{ .From }}
+
+LABEL \
+  org.label-schema.schema-version="1.0" \
+  org.label-schema.vendor="Elastic" \
```
Comment (andrewkroh) [inline dev-tools/packaging/templates/docker/Dockerfile.tmpl]: Community beats may want to reuse of image builders so this should be templated.

Comment (jsoriano) [reply dev-tools/packaging/templates/docker/Dockerfile.tmpl]: Parameterized now.

File: dev-tools/packaging/templates/docker/Dockerfile.tmpl [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+FROM {{ .from }}
+
+LABEL \
+  org.label-schema.schema-version="1.0" \
+  org.label-schema.vendor="{{ .BeatVendor }}" \
+  org.label-schema.name="{{ .BeatName }}" \
+  org.label-schema.version="{{ beat_version }}" \
+  org.label-schema.url="{{ .BeatURL }}" \
+  org.label-schema.vcs-url="{{ $repoInfo.RootImportPath }}" \
```
Comment (andrewkroh) [inline dev-tools/packaging/templates/docker/Dockerfile.tmpl]: I think adding org.label-schema.vcs-ref would be nice as well.

Comment (jsoriano) [reply dev-tools/packaging/templates/docker/Dockerfile.tmpl]: Added.

**Comment by jsoriano** [comment]
@andrewkroh @graphaelli some sanity tests are still pending, but I think this is ready for a first review, it'd be nice if you could give it another look. Thanks!

File: dev-tools/packaging/packages.yml [hunk 3]
```diff
@@ -157,6 +157,19 @@ shared:
         template: '{{ elastic_beats_dir }}/dev-tools/packaging/templates/windows/uninstall-service.ps1.tmpl'
         mode: 0755
 
+  - &docker_spec
+    <<: *binary_spec
+    extra_vars:
+      from: 'centos:7'
+      user: '{{ .BeatName }}'
+      repository: 'docker.elastic.co/beats'
```
Comment (graphaelli) [inline dev-tools/packaging/packages.yml]: can we add a param for this?

File: dev-tools/mage/dockerbuilder.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+	// Save the container as artifact
+	outputFile := b.OutputFile
+	if outputFile == "" {
+		outputTar, err := b.Expand(defaultBinaryName + ".docker.tar")
+		if err != nil {
+			return err
+		}
+		outputFile = filepath.Join(distributionsDir, outputTar)
+	}
+	if err := sh.Run("docker", "save", "-o", outputFile, tag); err != nil {
```
Comment (graphaelli) [inline dev-tools/mage/dockerbuilder.go]: I like this but it doesn't appear to be required for releases - the publisher currently does this + compression

File: dev-tools/packaging/packages.yml [hunk 5]
```diff
@@ -157,6 +157,19 @@ shared:
...
+      linux_capabilities: ''
```
Comment (graphaelli) [inline dev-tools/packaging/packages.yml]: feature request: something like `expose_port` - and then use like [this](https://github.com/elastic/apm-server-docker/blob/b1b58e119edd106eb2882fb5e3db6e2b9d9ef734/templates/Dockerfile.j2#L37)

File: dev-tools/mage/dockerbuilder.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+package mage
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/magefile/mage/sh"
+
```
Comment (graphaelli) [inline dev-tools/mage/dockerbuilder.go]: new `goimports` doesn't like this space

File: dev-tools/packaging/packages.yml (hunk 3 above)
Comment (jsoriano) [reply dev-tools/packaging/packages.yml]: As this is specific of docker packaging I left it as an extra var (like the specific extra vars for DMG builds). This extra var is optional, and can be set from `packages.yml` or overriden with `spec.ExtraVar()` from the magefile.

I am open to move it to other kind of parametrization though.

File: dev-tools/mage/dockerbuilder.go (hunk 4 above)
Comment (jsoriano) [reply dev-tools/mage/dockerbuilder.go]: Oh, I though we also needed to generate the artifact and publisher does `docker save` and `docker push`.
I am using it also for sanity checks, so I'd leave it if it is not a problem.

File: dev-tools/packaging/packages.yml (hunk 5 above)
Comment (jsoriano) [reply dev-tools/packaging/packages.yml]: Done, `expose_ports` can be a comma separated list of ports.

File: dev-tools/packaging/packages.yml (hunk 3 above)
Comment (jsoriano) [reply dev-tools/packaging/packages.yml]: I have added 8e6a32356 so it is not set by default on community beats.

File: dev-tools/mage/dockerbuilder.go (hunk 4 above)
Comment (graphaelli) [reply dev-tools/mage/dockerbuilder.go]: Not a problem from my perspective, agreed on keeping it.

File: .gitignore [hunk 7]
```diff
@@ -8,7 +8,7 @@
 /*/fields.yml
 /*/*.template*.json
 **/html_docs
-/*/_meta/kibana.generated
+/**/_meta/kibana.generated
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply .gitignore]: Ok, I'll remove the change from here.

File: dev-tools/packaging/package_test.go [hunk 8]
```diff
@@ -311,6 +332,30 @@ func checkModules(t *testing.T, name, prefix string, r *regexp.Regexp, p *packag
...
+			entrypoint := strings.TrimPrefix(entrypoint, "/")
+			entry, found := p.Contents[entrypoint]
+			if !found {
+				t.Fatalf("%s entrypoint not found in docker", entrypoint)
+			}
+			if mode := entry.Mode.Perm(); mode != expectedMode {
+				t.Fatalf("%s entrypoint mode is %s, expected: %s", entrypoint, mode, expectedMode)
+			}
+		} else {
+			t.Skip("TODO: check if binary is in $PATH")
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply dev-tools/packaging/package_test.go]: Ok, I'll change it to `t.Fatal`.

**Comment by jsoriano** [comment]
@andrewkroh thanks for the review, I have addressed your comments.

Regarding additional tests, it could be possible, we could maybe modify system tests to use docker, so the docker image is also tested with them, but I don't have any plan for it at the moment.

**Pull request merged**